xapian-applications/omega/clickmodel/simplifieddbn.cc

```diff
@@ -0,0 +1,213 @@
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <config.h>
+
+#include "simplifieddbn.h"
+
+#include <algorithm>
+#include <iomanip>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <string>
+#include <vector>
+
+using namespace std;
+
+#define QID 0
+#define DOCIDS 1
+#define CLICKS 2
+
+string
+SimplifiedDBN::name()
+{
+    return "SimplifiedDBN";
+}
+
+static vector<string>
+get_docid_list(const vector<vector<string>> &sessions)
+{
+    vector<string> docids;
+    string docid;
+    for (size_t i = 0; i < sessions.size(); ++i) {
+	for (size_t j = 0; j <= sessions[i][DOCIDS].length(); ++j) {
+	    if (sessions[i][DOCIDS][j] != ',' && sessions[i][1][j] != '\0') {
+		docid += sessions[i][DOCIDS][j];
+	    } else {
+		docids.push_back(docid);
+		docid.clear();
+	    }
+	}
+    }
+    return docids;
+}
+
+static vector<int>
+get_click_list(const vector<vector<string>> &sessions)
+{
+    vector<int> clicks;
+    string clickstring, clickcount;
+    int click = 0;
+    for (size_t i = 0; i < sessions.size(); ++i) {
+	for (size_t j = 0; j <= sessions[i][CLICKS].length(); ++j) {
+	    if (sessions[i][CLICKS][j] == '\0') {
+		clicks.push_back(click);
+	    } else if (sessions[i][CLICKS][j] != ',') {
+		// Get clickstring of the form "docid:click_count".
+		clickstring += sessions[i][CLICKS][j];
+
+		// Get click count string.
+		int delimiter_pos = clickstring.find(':');
+		clickcount = clickstring.substr(delimiter_pos + 1,
+			     clickstring.length() - 1 - delimiter_pos);
+
+		// Convert click count string to an integer.
+		stringstream ss(clickcount);
+		ss >> click;
+	    } else {
+		clickstring.clear();
+		clicks.push_back(click);
+	    }
+	}
+    }
+    return clicks;
+}
+
+vector<vector<string>>
+SimplifiedDBN::build_sessions(string logfile)
+{
+    ifstream file;
+    file.open(logfile, ios::in);
+
+    string line;
+
+    // Skip the first line.
+    getline(file, line);
+
+    vector<vector<string>> sessions;
+    vector<string> session_entry;
+
+    // Start reading file from the second line.
+    while (getline(file, line)) {
+
+	istringstream ss(line);
+
+	vector<string> row_data;
+
+	while (ss >> std::ws) {
+	    string column_element;
+	    if (ss.peek() == '"') {
+		int pos = ss.tellg();
+		ss.seekg(pos + 1);
+		char ch;
+		while (ss.get(ch)) {
+		    if (ch == '"')
+			break;
+		    column_element += ch;
+		}
+	    } else {
+		getline(ss, column_element, ',');
+	    }
+	    row_data.push_back(column_element);
+	}
+
+	string qid = row_data[0];
+	string query = row_data[1];
+	string docids = row_data[2];
+	string clicks = row_data[4];
+
+	session_entry.push_back(qid);
+	session_entry.push_back(docids);
+	session_entry.push_back(clicks);
+
+	sessions.push_back(session_entry);
+
+	session_entry.clear();
+    }
+
+    file.close();
+    return sessions;
+}
+
+void
+SimplifiedDBN::train(vector<vector<string>> &sessions)
```

Comment:
Shouldn't `sessions` be `const`?  It doesn't seem like it should be modified by this function.